### PR TITLE
Add support for Workflow Steps from Apps

### PR DESCRIPTION
**Pull request by misscoded**
###  Summary

This pull request contains changes outlined to incorporate support for [Workflow Steps from Apps](https://api.slack.com/workflows/steps).

---

> Workflow Steps provide the ability for Slack apps to create and process custom Workflow Steps for use in Workflows. Steps can be built to do things such as send data to external services, create tasks in project management systems, or update tickets. These steps can then be shared and used by anyone in the Workflow Builder to incorporate into Workflows they're building.
> 
> For more information on Workflow Steps, read full details here: https://api.slack.com/workflows/steps

### Category (place an `x` in each of the `[ ]`)

* [x] `slack_bolt.App` and/or its core components
* [x] `slack_bolt.async_app.AsyncApp` and/or its core components
* [ ] Adapters in `slack_bolt.adapter`
* [x] Document pages under `/docs`
* [x] Others

## Requirements (place an `x` in each `[ ]`)

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [ ] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

File: docs/_steps/adding_editing_workflow_step.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        "label": {
+          "type": "plain_text",
+          "text": "Task description",
+        },
+      },
+    ]
+
+    configure(blocks=blocks)
+  
+ws = WorkflowStep(callback_id="add_task", config={
```
Comment (seratch) [inline docs/_steps/adding_editing_workflow_step.md:61]: Just an idea but providing a decorator as below may be worth considering if we can provide both 1: passing a dict for config and 2: decorator for edit/update/execute listeners. I'm happy to work on this as necessary.

```python
add_task_ws = WorkflowStep(callback_id="add_task")

@add_task_ws.edit
def add_task_edit(ack, step, configure):
    pass
@add_task_ws.update
def add_task_update(ack, step):
    pass
@add_task_ws.execute
def add_task_execute():
    pass

app.step(add_task_ws)
```

File: docs/_steps/saving_workflow_step.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+To learn more about how to structure these parameters, [read the documentation](https://api.slack.com/reference/workflows/workflow_step).
+
+</div>
+
+```python
+def save_handler(ack, view, update):
+    ack()
+
+    values = view.state
```
Comment (seratch) [inline docs/_steps/saving_workflow_step.md:27]:
```suggestion
    values = view["state"]["values"]
```

File: docs/_steps/saving_workflow_step.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    task_name = values["task_name_input"]["name"]
+    task_description = values["task_description_input"]["description"]
+                
+    inputs = {
+      "task_name": {
```
Comment (seratch) [inline docs/_steps/saving_workflow_step.md:32]: Let's make the indents consistent (four spaces)

File: docs/_steps/saving_workflow_step.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+def save_handler(ack, view, update):
+    ack()
+
+    values = view.state
+    task_name = values["task_name_input"]["name"]
+    task_description = values["task_description_input"]["description"]
+                
+    inputs = {
+      "task_name": { 
+        "value": task_name.value
```
Comment (seratch) [inline docs/_steps/saving_workflow_step.md:33]: In Bolt for Python, payloads are always a dict.

```suggestion
        "value": task_name["value"]
```

File: docs/_steps/saving_workflow_step.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    values = view.state
+    task_name = values["task_name_input"]["name"]
+    task_description = values["task_description_input"]["description"]
+                
+    inputs = {
+      "task_name": { 
+        "value": task_name.value 
+      },
+      "task_description": { 
+        "value": task_description.value
```
Comment (seratch) [inline docs/_steps/saving_workflow_step.md:36]: same as above

File: slack_bolt/workflows/step/step.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    return CustomMiddleware(app_name=__name__, func=save_listener_middleware)
+
+
+#######################
+# Execute
+#######################
+
+
+def _build_execute_listener_middleware():
+    def save_listener_middleware(
```
Comment (misscoded) [inline slack_bolt/workflows/step/step.py:126]: I assume this is just a copy/paste and should be `execute_listener_middleware`.

**Review by misscoded — COMMENTED** [review]
Comments to get some clarity from @seratch before making any further changes.

File: samples/steps_from_apps.py [hunk 7]
```diff
@@ -19,156 +21,158 @@
...
+                "title": {"type": "plain_text", "text": "Your tasks!"},
+                "blocks": blocks,
+            },
+        )
+    except Exception as err:
+        fail(error={
+            "message": "Something wrong!"
+        })
+
+app.step(
```
[the remark this answers is not in the document]
Comment (seratch) [reply samples/steps_from_apps.py:170]: I added the kwargs pattern just for convenience and simplicity from a user perspective. It's also possible to use a class as below:

```python
app.step(WorkflowStep(
    callback_id="copy_review",
    edit=edit,
    save=save,
    execute=execute,
))
```

Feel free to remove the kwargs pattern if you don't see the necessity to have it. I don't have any strong reason to add it.

File: slack_bolt/app/app.py [hunk 8]
```diff
@@ -335,6 +336,34 @@ def middleware(self, *args) -> None:
                     CustomMiddleware(app_name=self.name, func=middleware_or_callable)
                 )
 
+    # -------------------------
+    # Workflows: Steps from Apps
+
+    def step(
+        self,
+        callback_id: Union[str, Pattern, WorkflowStep],
+        save_callback_id: Optional[Union[str, Pattern]] = None,
```
[the remark this answers is not in the document]
Comment (seratch) [reply slack_bolt/app/app.py:345]: I also agree it's totally fine not to have the option and it should work in most cases.

If we think about a possible reason to have this option, let's say a developer doesn't use `configure` for some reason (I'm not sure what it is though). In this case, the developer may want to have an option to set this. But again I'm not sure why the dev has to do so.

File: slack_bolt/app/app.py [hunk 9]
```diff
@@ -335,6 +336,34 @@ def middleware(self, *args) -> None:
                     CustomMiddleware(app_name=self.name, func=middleware_or_callable)
                 )
 
+    # -------------------------
+    # Workflows: Steps from Apps
+
+    def step(
+        self,
+        callback_id: Union[str, Pattern, WorkflowStep],
```
Comment (seratch) [inline slack_bolt/app/app.py:344]: So, if you remove the kwargs pattern here, this method can accept only a `WorkflowStep` object as an argument.

File: slack_bolt/app/app.py [hunk 10]
```diff
@@ -335,6 +336,34 @@ def middleware(self, *args) -> None:
...
+        save_callback_id: Optional[Union[str, Pattern]] = None,
+        edit: Optional[Union[Callable[..., Optional[BoltResponse]], Listener]] = None,
+        save: Optional[Union[Callable[..., Optional[BoltResponse]], Listener]] = None,
+        execute: Optional[
+            Union[Callable[..., Optional[BoltResponse]], Listener]
+        ] = None,
+    ):
+        """Registers a new Workflow Steps from Apps listeners."""
```
Comment (seratch) [inline slack_bolt/app/app.py:352]: Sorry, please change this docstring. This is grammatically incorrect and doesn't make sense.

File: slack_bolt/workflows/step/step.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    callback_id: str
+    edit: Listener
+    save: Listener
+    execute: Listener
+
+    def __init__(
+        self,
+        *,
+        callback_id: str,
+        edit: Union[Callable[..., Optional[BoltResponse]], Listener],
```
[the remark this answers is not in the document]
Comment (seratch) [reply slack_bolt/workflows/step/step.py:29]: We're not going to change this. Instead, we want to provide a handy way to attach listener middleware to this. The `Listener` can have `middleware: List[Middleware]`. Developers should be able to attach them by using a new decorator.

```python
ws = WorkflowStep(callback_id="foo")

@ws.edit(middleware=[do_this, do_that])
def edit_foo(ack, configure):
    ack()
    configure(...)

app.use(ws) # We can validate the ws already have three listener functions when registering this to App
```

Or we may want to have another one like `WorkflowStep.builder(callback_id="foo")`

```python
ws = WorkflowStep.builder(callback_id="foo")

@ws.edit(middleware=[do_this, do_that])
def edit_foo(ack, configure):
    ack()
    configure(...)

app.use(ws.build()) # build method verifies if there are sufficient listeners in it.
```

File: slack_bolt/workflows/step/step.py (hunk 6 above)
Comment (seratch) [reply slack_bolt/workflows/step/step.py:126]: You're right. please correct it.

**Review by seratch — COMMENTED** [review]
Thanks for clarifying these! @misscoded

File: slack_bolt/app/app.py (hunk 10 above)
Comment (misscoded) [reply slack_bolt/app/app.py:352]: Changing to `"""Registers a new Workflow Step listener"""`

File: slack_bolt/workflows/step/step.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (misscoded) [reply slack_bolt/workflows/step/step.py:29]: In the spirit of fostering shared knowledge of our different implementations, @aoberoi you had the original opinion that steered us to support the user passing in an array of middleware to the configuration object in Bolt for JS. 

Just wanted to highlight how Bolt for Python's implementation will do this based on the discussion above.

File: samples/steps_from_apps.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (misscoded) [reply samples/steps_from_apps.py:170]: Based on this, I am going to change the examples in the docs to reflect the above, as I get the impression that it's the preferred way to instantiate a Step.

File: slack_bolt/workflows/step/utilities/async_update.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+class AsyncUpdate:
+    def __init__(self, *, client: AsyncWebClient, body: dict):
+        self.client = client
+        self.body = body
+
+    async def __call__(self, *, inputs: dict, outputs: list,) -> None:
+        await self.client.workflows_updateStep(
+            workflow_step_edit_id=self.body["workflow_step"]["workflow_step_edit_id"],
+            inputs=inputs,
```
Comment (misscoded) [inline slack_bolt/workflows/step/utilities/async_update.py:12]: TODO: `inputs` and `outputs` aren't required here. Also possible to pass in additional args

File: slack_bolt/workflows/step/utilities/update.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+class Update:
+    def __init__(self, *, client: WebClient, body: dict):
+        self.client = client
+        self.body = body
+
+    def __call__(self, *, inputs: dict, outputs: list,) -> None:
+        self.client.workflows_updateStep(
+            workflow_step_edit_id=self.body["workflow_step"]["workflow_step_edit_id"],
+            inputs=inputs,
```
Comment (misscoded) [inline slack_bolt/workflows/step/utilities/update.py:12]: TODO: `inputs` and `outputs` aren't required here. Also possible to pass in additional args

File: slack_bolt/workflows/step/utilities/complete.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    def __init__(self, *, client: WebClient, body: dict):
+        self.client = client
+        self.body = body
+
+    def __call__(self, *, outputs: dict,) -> None:
+        self.client.workflows_stepCompleted(
+            workflow_step_execute_id=self.body["event"]["workflow_step"][
+                "workflow_step_execute_id"
+            ],
+            outputs=outputs,
```
Comment (misscoded) [inline slack_bolt/workflows/step/utilities/complete.py:14]: TODO: `outputs` is not required

File: slack_bolt/workflows/step/utilities/async_complete.py [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+    def __init__(self, *, client: AsyncWebClient, body: dict):
+        self.client = client
+        self.body = body
+
+    async def __call__(self, *, outputs: dict,) -> None:
+        await self.client.workflows_stepCompleted(
+            workflow_step_execute_id=self.body["event"]["workflow_step"][
+                "workflow_step_execute_id"
+            ],
+            outputs=outputs,
```
Comment (misscoded) [inline slack_bolt/workflows/step/utilities/async_complete.py:14]: TODO: `outputs` is not required

File: docs/_steps/creating_workflow_step.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+When instantiating a new `WorkflowStep`, pass in the step's `callback_id` and a configuration object.
+
+The configuration object contains three keys: `edit`, `save`, and `execute`. Each of these keys must be a single callback or a list of callbacks. All callbacks have access to a `step` object that contains information about the workflow step event.
+
+After instantiating a `WorkflowStep`, you can pass it into `app.step()`. Behind the scenes, your app will listen and respond to the workflow step’s events using the callbacks provided in the configuration object.
+
+</div>
+
+```python
+from slack_bolt import App, WorkflowStep
```
Comment (misscoded) [inline docs/_steps/creating_workflow_step.md:21]: @seratch I want to verify that we want to adjust this import to be 

`from slack_bolt.workflows.step import WorkflowStep` 

and not expose it directly off of the package

Comment (seratch) [reply docs/_steps/creating_workflow_step.md:21]: @misscoded If you add only `WorkflowStep` to the top-level, I think it's acceptable as the class is part of the public APIs of the `App` class. I want to mention the same about `AsyncWorkflowSteps` in `slack_bolt.async_app`.

Comment (misscoded) [reply docs/_steps/creating_workflow_step.md:21]: We've agreed that adding `WorkflowStep` to the top-level breaks an existing pattern, so for now, we will choose not to do so.

**Review by seratch — APPROVED** [review]
Thank you very much for your huge efforts here and the great work! The change looks great to me. I will add more unit tests afterward.

**Pull request merged**